### PR TITLE
GH2333: Rewrite Cake CLI (RFC-0001)

### PR DESCRIPTION
**Pull request by patriksvensson**
RFC: [0001](https://github.com/cake-build/rfcs/blob/master/text/0001-client-rewrite.md)

## Notes

I'm putting the code temporarily in it's own project (Cake.Client) to avoid merge conflicts. Before merging, this project should replace the `Cake` project.

### Aliases

- [x] `IEnumerable<T> Arguments<T>(string name)`

### Features

- [x] Compilation
- [x] Debugging
- [x] Bootstrapping
- [x] Dry run
- [x] Tree
- [x] Descriptions
- [x] Version
- [x] Info

### Other

- [x] Module loading
- [x] Multiple arguments with the same name

### CODE REVIEW

File: src/Cake/Cake.csproj [hunk 1]
```diff
@@ -1,21 +1,23 @@
 ﻿<Project Sdk="Microsoft.NET.Sdk">
+
   <PropertyGroup>
-    <AssemblyName>Cake</AssemblyName>
-    <TargetFrameworks Condition=" '$(PackAsTool)' != 'true'">net461;netcoreapp2.0</TargetFrameworks>
     <OutputType>Exe</OutputType>
-    <PlatformTarget>AnyCpu</PlatformTarget>
```
Comment (patriksvensson) [inline src/Cake/Cake.csproj]: Keep `PlatformTarget`

File: src/Cake/Features/CakeFeature.cs [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+            _bootstrapper.Bootstrap(registrar, configuration, arguments);
+            action?.Invoke(registrar);
+
+            return registrar.BuildContainer();
+        }
+
+        protected ICakeConfiguration ReadConfiguration(IRemainingArguments remaining, FilePath script)
+        {
+            var provider = new CakeConfigurationProvider(_fileSystem, _environment);
+            var root = script.GetDirectory();
```
Comment (patriksvensson) [inline src/Cake/Features/CakeFeature.cs]: Get the directory as an argument

File: src/Cake/Infrastructure/Composition/ModuleLoader.cs [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+
+            if (Activator.CreateInstance(type, arguments) is ICakeModule module)
+            {
+                return module;
+            }
+
+            return null;
+        }
+
+        private static ConstructorInfo GetGreediestConstructor(Type type)
```
Comment (patriksvensson) [inline src/Cake/Infrastructure/Composition/ModuleLoader.cs]: Replace with suggestion from @Meir017

File: src/Cake/Infrastructure/ContainerBootstrapper.cs [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+using Cake.NuGet;
+using Spectre.Cli;
+
+namespace Cake.Client.Infrastructure
+{
+    /// <summary>
+    /// Responsible for registering all dependencies
+    /// that is required for executing a script.
+    /// </summary>
+    public interface IContainerBootstrapper
```
Comment (patriksvensson) [inline src/Cake/Infrastructure/ContainerBootstrapper.cs]: New name for this? Don't like to use bootstrapper in another meaning than bootstrapping Cake scripts.

File: src/Cake/Infrastructure/Diagnostics/CakeDebugger.cs [hunk 5]
```diff
@@ -0,0 +1,34 @@
+﻿using System;
+using System.Diagnostics;
+using System.Threading;
+using System.Threading.Tasks;
+using Cake.Core.Diagnostics;
+
+namespace Cake.Client.Infrastructure.Diagnostics
+{
+    public sealed class CakeDebugger : ICakeDebugger
```
Comment (patriksvensson) [inline src/Cake/Infrastructure/Diagnostics/CakeDebugger.cs]: Move this to Cake.Core?

File: src/Cake/Infrastructure/Diagnostics/ICakeDebugger.cs [hunk 6]
```diff
@@ -0,0 +1,9 @@
+﻿using System;
+
+namespace Cake.Client.Infrastructure.Diagnostics
+{
+    public interface ICakeDebugger
```
Comment (patriksvensson) [inline src/Cake/Infrastructure/Diagnostics/ICakeDebugger.cs]: Move this to Cake.Core?

File: src/Cake.Core/CakeArguments.cs [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+        /// <summary>
+        /// Gets all values for an argument.
+        /// </summary>
+        /// <param name="name">The argument name.</param>
+        /// <returns>The argument values.</returns>
+        public string[] GetArguments(string name)
+        {
+            return _arguments.Contains(name)
+                ? _arguments[name].ToArray()
+                : null;
```
Comment (patriksvensson) [inline src/Cake.Core/CakeArguments.cs]: Should we really return null here or an empty array?

File: src/Cake.Core/Graph/CakeGraph.cs [hunk 8]
```diff
@@ -8,21 +8,37 @@
 
 namespace Cake.Core.Graph
 {
-    internal sealed class CakeGraph
+    /// <summary>
+    /// Represents the Cake task graph.
+    /// </summary>
+    public sealed class CakeGraph
```
Comment (patriksvensson) [inline src/Cake.Core/Graph/CakeGraph.cs]: Made public due to the new tree functionality. Should we find another way of getting the task graph? Expose it by providing a read only interface to the tree script host somehow?

File: src/Cake.Core/ICakeArguments.cs [hunk 9]
```diff
@@ -19,10 +21,10 @@ public interface ICakeArguments
...
 
         /// <summary>
-        /// Gets an argument.
+        /// Gets all values for an argument.
         /// </summary>
         /// <param name="name">The argument name.</param>
-        /// <returns>The argument value.</returns>
-        string GetArgument(string name);
+        /// <returns>The argument values.</returns>
+        string[] GetArguments(string name);
```
Comment (patriksvensson) [inline src/Cake.Core/ICakeArguments.cs]: Return an ICollection<T> here instead?

File: src/Cake.Core/Properties/AssemblyInfo.cs [hunk 10]
```diff
@@ -22,6 +22,7 @@
 
 // Make internals visible to other Cake assemblies.
 [assembly: InternalsVisibleTo("Cake")]
+[assembly: InternalsVisibleTo("Cake.Client")]
```
Comment (patriksvensson) [inline src/Cake.Core/Properties/AssemblyInfo.cs]: Remove

File: src/Cake.Core/Properties/AssemblyInfo.cs [hunk 11]
```diff
@@ -22,6 +22,7 @@
 
 // Make internals visible to other Cake assemblies.
 [assembly: InternalsVisibleTo("Cake")]
```
Comment (patriksvensson) [inline src/Cake.Core/Properties/AssemblyInfo.cs]: Remove

**Comment by patriksvensson** [comment]
@mholo65 Reminder for you to take a look at where you want ScriptAssemblyResolver to be.

File: src/Cake/Program.cs [hunk 12]
```diff
@@ -1,112 +1,94 @@
...
+                // Top level examples.
+                config.AddExample(new[] { string.Empty });
+                config.AddExample(new[] { "build.cake", "--verbosity", "quiet" });
+                config.AddExample(new[] { "build.cake", "--tree" });
+            });
 
-                    // Parse the options.
-                    var parser = container.Resolve<IArgumentParser>();
-                    var options = parser.Parse(args);
+            return await app.RunAsync(args);
```
Comment (mholo65) [inline src/Cake/Program.cs]: @patriksvensson maybe `ScriptAssemblyResolver` could be added somewhere around here? Or better yet, before loading any third party modules.

**Comment by patriksvensson** [comment]
@cake-build/cake-team Anyone want to take a look at this?

File: build.cake [hunk 13]
```diff
@@ -189,7 +189,8 @@ Task("Run-Unit-Tests")
         // .NET Framework/Mono
         // Total hack, but gets the work done.
         var framework = "net46";
-        if(project.ToString().EndsWith("Cake.Tests.csproj")) {
+        if(project.ToString().EndsWith("Cake.Tests.csproj") ||
```
Comment (Meir017) [inline build.cake]: could this change to?
```cs
project.ToString().EndsWith(".Tests.csproj")
```

File: build.cake [hunk 14]
```diff
@@ -232,14 +233,11 @@ Task("Copy-Files")
...
     CopyFileToDirectory("./LICENSE", parameters.Paths.Directories.ArtifactsBinNetCore);
-
-    // Copy Cake.XML (since publish does not do this anymore)
-    CopyFileToDirectory("./src/Cake/bin/" + parameters.Configuration + "/net461/Cake.xml", parameters.Paths.Directories.ArtifactsBinFullFx);
-    CopyFileToDirectory("./src/Cake/bin/" + parameters.Configuration + "/netcoreapp2.0/Cake.xml", parameters.Paths.Directories.ArtifactsBinNetCore);
 });
 
 Task("Validate-Version")
     .IsDependentOn("Copy-Files")
+    .WithCriteria(false)
```
Comment (Meir017) [inline build.cake]: can this task run?

File: src/Cake.Core/Scripting/ScriptRunner.cs [hunk 15]
```diff
@@ -52,61 +52,23 @@ public sealed class ScriptRunner : IScriptRunner
...
         /// <summary>
         /// Runs the script using the specified script host.
         /// </summary>
         /// <param name="host">The script host.</param>
         /// <param name="scriptPath">The script.</param>
-        /// <param name="arguments">The arguments.</param>
-        public void Run(IScriptHost host, FilePath scriptPath, IDictionary<string, string> arguments)
+        public void Run(IScriptHost host, FilePath scriptPath)
         {
             if (host == null)
```
Comment (Meir017) [inline src/Cake.Core/Scripting/ScriptRunner.cs]: use throw expression here? same as in the ctor

File: src/Cake/Features/Bootstrapping/BootstrapFeature.cs [hunk 16]
```diff
@@ -0,0 +1,70 @@
...
+            ICakeEnvironment environment,
+            IContainerBootstrapper bootstrapper) : base(fileSystem, environment, bootstrapper)
+        {
+            _environment = environment;
+        }
+
+        public int Run(IRemainingArguments arguments, BootstrapFeatureSettings settings)
+        {
+            // Fix the script path.
+            settings.Script = settings.Script ?? new FilePath("build.cake");
```
Comment (Meir017) [inline src/Cake/Features/Bootstrapping/BootstrapFeature.cs]: maybe move the `new FilePath("build.cake")` to a constant or set a default value to the `BootstrapFeatureSettings .Script` property?

File: src/Cake/Infrastructure/Composition/ModuleLoader.cs (hunk 3 above)
Comment (patriksvensson) [reply src/Cake/Infrastructure/Composition/ModuleLoader.cs]: @Meir017 Remeber what your suggestion was here? I can't seem to find it 😢

**Comment by patriksvensson** [comment]
@cake-build/cake-team Fixed stuff, squashed commits and rebased against develop. Could you guys take a look?

File: src/Cake/Infrastructure/Composition/ModuleLoader.cs (hunk 3 above)
Comment (Meir017) [reply src/Cake/Infrastructure/Composition/ModuleLoader.cs]: I think this:
```cs
private static ConstructorInfo GetGreediestConstructor(Type type) => type.GetTypeInfo()
    .GetConstructors()
    .OrderByDescending(ctor => ctor.GetParameters().Length)
    .First();
```

File: src/Cake/Infrastructure/ContainerBootstrapper.cs (hunk 4 above)
Comment (patriksvensson) [reply src/Cake/Infrastructure/ContainerBootstrapper.cs]: I'm renaming this to `IContainerConfigurator`.

File: src/Cake.Common/ArgumentAliases.cs [hunk 17]
```diff
@@ -57,7 +59,7 @@ public static bool HasArgument(this ICakeContext context, string name)
         /// <returns>The value of the argument.</returns>
         /// <example>
         /// <code>
-        /// //Cake.exe .\argument.cake -myArgument="is valid" -loopCount = 5
+        /// //Cake.exe .\argument.cake --myArgument="is valid" -loopCount=5
```
Comment (devlead) [inline src/Cake.Common/ArgumentAliases.cs:62]: `-loopCount=5` should be `--loopCount=5` I'll fix.

File: src/Cake.Common/ArgumentAliases.cs [hunk 18]
```diff
@@ -96,7 +165,7 @@ public static T Argument<T>(this ICakeContext context, string name)
         /// <returns>The value of the argument if it exist; otherwise <paramref name="defaultValue"/>.</returns>
         /// <example>
         /// <code>
-        /// //Cake.exe .\argument.cake -myArgument="is valid" -loopCount = 5
+        /// //Cake.exe .\argument.cake --myArgument="is valid" -loopCount=5
```
Comment (devlead) [inline src/Cake.Common/ArgumentAliases.cs:168]: `-loopCount=5` should be `--loopCount=5` I'll fix.

**Comment by devlead** [comment]
@cake-build/cake-team I've rebased this one against latest develop and fixed a few stylecop issues and merge conflicts with package references.
It should be ready for review now.

**Comment by ecampidoglio** [comment]
The [original RFC](https://github.com/cake-build/rfcs/blob/master/text/0001-client-rewrite.md) proposed the introduction of _verbs_ to trigger certain behaviors. For example:

```sh
cake bootstrap
```

instead of

```sh
cake --bootstrap
```

Has this idea been dropped? It doesn't seem to be part of this PR.

**Comment by gep13** [comment]
@ecampidoglio my understanding was that with this rewrite, there would now be an implicit bootstrapping performed, without the need to trigger anything.  I haven't had a chance to take this for a spin yet though.

**Comment by ecampidoglio** [comment]
@gep13 Nice. That would mean we can remove the `bootstrap` option from the [cake-action](https://github.com/cake-build/cake-action) at some point.

Regarding the verbs, the RFC [described the introduction of a `build` command](https://github.com/cake-build/rfcs/blob/master/text/0001-client-rewrite.md#motivation), similar to `dotnet build`:

> It would make it easier to use Cake if we introduced proper commands and subcommands like most people are used to from applications such as `git` and `dotnet`. Of course we would need to be backwards compatible with how Cake works today (without commands and sub commands) but this functionality could be phased out before a 1.0 release of Cake. The `build` command should however be the default command.

However, I couldn't find any traces of this in the PR. Has the idea been scrapped or postponed?

**Comment by gep13** [comment]
@ecampidoglio hmm, not sure on that.  That would be a question for @patriksvensson

**Comment by patriksvensson** [comment]
@ecampidoglio You're absolutely right. I haven't added it since we haven't had a proper discussion about what commands we should expose from the root.

**Comment by devlead** [comment]
@cake-build/cake-team I've rebased this one against latest develop.
Please review.

File: src/Cake/Features/Introspection/VersionFeature.cs [hunk 19]
```diff
@@ -0,0 +1,27 @@
+﻿using Cake.Core;
```
Comment (gep13) [inline src/Cake/Features/Introspection/VersionFeature.cs:1]: This is missing license header

File: src/Cake/Features/Introspection/VersionResolver.cs [hunk 20]
```diff
@@ -0,0 +1,26 @@
+﻿using System.Diagnostics;
```
Comment (gep13) [inline src/Cake/Features/Introspection/VersionResolver.cs:1]: This is missing license header

File: src/Cake/Infrastructure/Composition/AutofacTypeRegistrar.cs [hunk 21]
```diff
@@ -0,0 +1,56 @@
+﻿using System;
```
Comment (gep13) [inline src/Cake/Infrastructure/Composition/AutofacTypeRegistrar.cs:1]: This is missing license header

File: src/Cake/Infrastructure/Composition/AutofacTypeResolver.cs [hunk 22]
```diff
@@ -0,0 +1,21 @@
+﻿using System;
```
Comment (gep13) [inline src/Cake/Infrastructure/Composition/AutofacTypeResolver.cs:1]: This is missing license header

File: src/Cake/Infrastructure/Composition/ContainerRegistrationBuilder.cs [hunk 23]
```diff
@@ -1,16 +1,12 @@
-﻿// Licensed to the .NET Foundation under one or more agreements.
```
Comment (gep13) [inline src/Cake/Infrastructure/Composition/ContainerRegistrationBuilder.cs:1]: Is there a reason that the license header was removed here?

File: src/Cake/Infrastructure/Composition/ModuleLoader.cs [hunk 24]
```diff
@@ -0,0 +1,64 @@
+﻿using System;
```
Comment (gep13) [inline src/Cake/Infrastructure/Composition/ModuleLoader.cs:1]: This is missing license header

File: src/Cake/Infrastructure/Composition/ModuleSearcher.cs [hunk 25]
```diff
@@ -1,54 +1,52 @@
-﻿// Licensed to the .NET Foundation under one or more agreements.
```
Comment (gep13) [inline src/Cake/Infrastructure/Composition/ModuleSearcher.cs:1]: Is there a reason that the license header was removed here?

File: src/Cake/Infrastructure/ContainerConfigurator.cs [hunk 26]
```diff
@@ -0,0 +1,59 @@
+﻿using Cake.Common.Modules;
```
Comment (gep13) [inline src/Cake/Infrastructure/ContainerConfigurator.cs:1]: This is missing license header

File: src/Cake/Infrastructure/ContainerConfigurator.cs [hunk 27]
```diff
@@ -0,0 +1,59 @@
...
+using Cake.NuGet;
+using Spectre.Cli;
+
+namespace Cake.Infrastructure
+{
+    /// <summary>
+    /// Responsible for registering all dependencies
+    /// that is required for executing a script.
+    /// </summary>
+    public interface IContainerConfigurator
```
Comment (gep13) [inline src/Cake/Infrastructure/ContainerConfigurator.cs:19]: Should this be in its own file?

File: src/Cake/Infrastructure/Converters/FilePathConverter.cs [hunk 28]
```diff
@@ -0,0 +1,19 @@
+﻿using System;
```
Comment (gep13) [inline src/Cake/Infrastructure/Converters/FilePathConverter.cs:1]: This is missing license header

File: src/Cake/Infrastructure/Converters/VerbosityConverter.cs [hunk 29]
```diff
@@ -0,0 +1,47 @@
+﻿using System;
```
Comment (gep13) [inline src/Cake/Infrastructure/Converters/VerbosityConverter.cs:1]: This is missing license header

File: src/Cake/Infrastructure/IScriptHostSettings.cs [hunk 30]
```diff
@@ -0,0 +1,13 @@
+﻿using System;
```
Comment (gep13) [inline src/Cake/Infrastructure/IScriptHostSettings.cs:1]: This is missing license header

File: src/Cake/Program.cs [hunk 31]
```diff
@@ -1,115 +1,93 @@
-﻿// Licensed to the .NET Foundation under one or more agreements.
```
Comment (gep13) [inline src/Cake/Program.cs:1]: Is there a reason that the license header was removed here?

File: src/Cake/Features/Introspection/InfoFeature.cs [hunk 32]
```diff
@@ -1,30 +1,28 @@
-// Licensed to the .NET Foundation under one or more agreements.
```
Comment (gep13) [inline src/Cake/Features/Introspection/InfoFeature.cs:1]: Is there a reason that the license header was removed here?

File: src/Cake/Features/CakeFeature.cs [hunk 33]
```diff
@@ -0,0 +1,52 @@
+﻿using System;
```
Comment (gep13) [inline src/Cake/Features/CakeFeature.cs:1]: This is missing license header

File: src/Cake/Features/Building/BuildHostKind.cs [hunk 34]
```diff
@@ -0,0 +1,10 @@
+﻿namespace Cake.Features.Building
```
Comment (gep13) [inline src/Cake/Features/Building/BuildHostKind.cs:1]: This is missing license header

File: src/Cake/Features/Building/BuildFeatureSettings.cs [hunk 35]
```diff
@@ -0,0 +1,20 @@
+﻿using Cake.Core.Diagnostics;
```
Comment (gep13) [inline src/Cake/Features/Building/BuildFeatureSettings.cs:1]: This is missing license header

File: src/Cake/Features/Building/BuildFeature.cs [hunk 36]
```diff
@@ -0,0 +1,131 @@
+﻿using System;
```
Comment (gep13) [inline src/Cake/Features/Building/BuildFeature.cs:1]: This is missing license header

File: src/Cake/Features/Bootstrapping/BootstrapFeatureSettings.cs [hunk 37]
```diff
@@ -0,0 +1,11 @@
+﻿using Cake.Core.Diagnostics;
```
Comment (gep13) [inline src/Cake/Features/Bootstrapping/BootstrapFeatureSettings.cs:1]: This is missing license header

File: src/Cake/Features/Bootstrapping/BootstrapFeature.cs [hunk 38]
```diff
@@ -0,0 +1,70 @@
+﻿using System;
```
Comment (gep13) [inline src/Cake/Features/Bootstrapping/BootstrapFeature.cs:1]: This is missing license header

File: src/Cake/Commands/DefaultCommandSettings.cs [hunk 39]
```diff
@@ -0,0 +1,55 @@
+﻿using System.ComponentModel;
```
Comment (gep13) [inline src/Cake/Commands/DefaultCommandSettings.cs:1]: This is missing license header

File: src/Cake/Commands/DefaultCommand.cs [hunk 40]
```diff
@@ -0,0 +1,83 @@
+﻿using Cake.Core.Diagnostics;
```
Comment (gep13) [inline src/Cake/Commands/DefaultCommand.cs:1]: This is missing license header

File: src/Cake.Tests/Utilities/TestContainerConfigurator.cs [hunk 41]
```diff
@@ -0,0 +1,39 @@
+﻿using System;
```
Comment (gep13) [inline src/Cake.Tests/Utilities/TestContainerConfigurator.cs:1]: This is missing license header

File: src/Cake.Tests/Unit/ProgramTests.cs [hunk 42]
```diff
@@ -0,0 +1,131 @@
+using System.Threading.Tasks;
```
Comment (gep13) [inline src/Cake.Tests/Unit/ProgramTests.cs:1]: This is missing license header

File: src/Cake.Tests/Unit/Features/VersionFeatureTests.cs [hunk 43]
```diff
@@ -0,0 +1,40 @@
+﻿using Cake.Features.Introspection;
```
Comment (gep13) [inline src/Cake.Tests/Unit/Features/VersionFeatureTests.cs:1]: This is missing license header

File: src/Cake.Tests/Unit/Features/InfoFeatureTests.cs [hunk 44]
```diff
@@ -0,0 +1,41 @@
+﻿using Cake.Features.Introspection;
```
Comment (gep13) [inline src/Cake.Tests/Unit/Features/InfoFeatureTests.cs:1]: This is missing license header

File: src/Cake.Tests/Unit/Features/BuildFeatureTests.cs [hunk 45]
```diff
@@ -0,0 +1,59 @@
+﻿using System;
```
Comment (gep13) [inline src/Cake.Tests/Unit/Features/BuildFeatureTests.cs:1]: This is missing license header

File: src/Cake.Tests/Unit/Features/BootstrapFeatureTests.cs [hunk 46]
```diff
@@ -0,0 +1,30 @@
+﻿using System.Collections.Generic;
```
Comment (gep13) [inline src/Cake.Tests/Unit/Features/BootstrapFeatureTests.cs:1]: This is missing license header

File: src/Cake.Tests/Fixtures/ProgramFixtureResult.cs [hunk 47]
```diff
@@ -0,0 +1,7 @@
+﻿namespace Cake.Tests.Fixtures
```
Comment (gep13) [inline src/Cake.Tests/Fixtures/ProgramFixtureResult.cs:1]: This is missing license header

File: src/Cake.Tests/Fixtures/ProgramFixture.cs [hunk 48]
```diff
@@ -0,0 +1,67 @@
+﻿using System;
```
Comment (gep13) [inline src/Cake.Tests/Fixtures/ProgramFixture.cs:1]: This is missing license header

File: src/Cake.Tests/Fixtures/BuildFeatureFixtureResult.cs [hunk 49]
```diff
@@ -0,0 +1,11 @@
+﻿using Cake.Core.Scripting;
```
Comment (gep13) [inline src/Cake.Tests/Fixtures/BuildFeatureFixtureResult.cs:1]: This is missing license header

File: src/Cake.Tests/Fixtures/BuildFeatureFixture.cs [hunk 50]
```diff
@@ -0,0 +1,97 @@
+﻿using System.Collections.Generic;
```
Comment (gep13) [inline src/Cake.Tests/Fixtures/BuildFeatureFixture.cs:1]: This is missing license header

File: src/Cake.Tests/Fakes/FakeVersionResolver.cs [hunk 51]
```diff
@@ -0,0 +1,26 @@
+﻿using Cake.Features.Introspection;
```
Comment (gep13) [inline src/Cake.Tests/Fakes/FakeVersionResolver.cs:1]: This is missing license header

File: src/Cake.Tests/Fakes/FakeScriptSession.cs [hunk 52]
```diff
@@ -0,0 +1,34 @@
+﻿using System;
```
Comment (gep13) [inline src/Cake.Tests/Fakes/FakeScriptSession.cs:1]: This is missing license header

File: src/Cake.Tests/Fakes/FakeScriptEngine.cs [hunk 53]
```diff
@@ -0,0 +1,21 @@
+﻿using Cake.Core.Scripting;
```
Comment (gep13) [inline src/Cake.Tests/Fakes/FakeScriptEngine.cs:1]: This is missing license header

File: src/Cake.Tests/Fakes/FakeRemainingArguments.cs [hunk 54]
```diff
@@ -0,0 +1,20 @@
+﻿using System;
```
Comment (gep13) [inline src/Cake.Tests/Fakes/FakeRemainingArguments.cs:1]: This is missing license header

File: src/Cake.Tests/Fakes/FakeDebugger.cs [hunk 55]
```diff
@@ -0,0 +1,15 @@
+﻿using System;
```
Comment (gep13) [inline src/Cake.Tests/Fakes/FakeDebugger.cs:1]: This is missing license header

File: src/Cake.Core/Diagnostics/ICakeDebugger.cs [hunk 56]
```diff
@@ -0,0 +1,16 @@
+﻿using System;
```
Comment (gep13) [inline src/Cake.Core/Diagnostics/ICakeDebugger.cs:1]: This is missing license header

File: src/Cake.Core/Diagnostics/CakeDebugger.cs [hunk 57]
```diff
@@ -0,0 +1,33 @@
+﻿using System;
```
Comment (gep13) [inline src/Cake.Core/Diagnostics/CakeDebugger.cs:1]: This is missing license header

File: src/Cake.Core/CakeArguments.cs [hunk 58]
```diff
@@ -0,0 +1,46 @@
+﻿using System.Collections.Generic;
```
Comment (gep13) [inline src/Cake.Core/CakeArguments.cs:1]: This is missing license header

**Comment by gep13** [comment]
@patriksvensson your changes have been merged, thanks for your contribution 👍

**Pull request merged**